### PR TITLE
feat: add predicate support baseline with tests

### LINKED ISSUE #341: Support for predicates
The Fuel VM allows for a Coin or Message input to be locked behind a predicate. To spend one of these inputs, the predicate is provided as a field of the input. If `hash(predicate) == Input.owner && eval(predicate)` , the UTXO can be spent. Note that in this case, no signature is required as the predicate alone is sufficient to spend the UTXO.

The SDK will therefore need to be able to:
- attach a compiled predicate to an Input.
- attach predicate data to an input. (This is analagous to providing scriptdata to a script - the data are parameters to the main function)
- send a transaction without the need for any wallet or signer
- set an output's `owner` to the hash of a compiled predicate

### PR DESCRIPTION
**Pull request by camsjams**
Closes #341 

Will be enhanced once https://github.com/FuelLabs/sway/issues/1981 is completed

PR for Sway is ready for review: https://github.com/FuelLabs/sway/pull/1984

### CODE REVIEW

File: packages/fuels-abigen-macro/tests/harness.rs
Comment (vnepveu) [inline packages/fuels-abigen-macro/tests/harness.rs:1928]: Here I feel like you can definitely "group" tests together, because there seems to be a lot of boilerplate code being repeated.

File: packages/fuels-abigen-macro/tests/harness.rs
Comment (vnepveu) [inline packages/fuels-abigen-macro/tests/harness.rs:1958]: You can use just `connected_provider.get_balance(asset_id)` iirc

File: packages/fuels-abigen-macro/tests/harness.rs
Comment (vnepveu) [inline packages/fuels-abigen-macro/tests/harness.rs:1969]: Use `?` here too, even in tests. For this, you will have to change the return type of the test to `Result<(), Error>`, see examples above in this file.

**Review by vnepveu — CHANGES_REQUESTED** [review]
Good first shot! My review isn't as thorough as it could be because I think I gave you enough material to improve for now :smiley_cat:. Congrats on this first Rust PR!

File: packages/fuels-abigen-macro/tests/test_projects/predicate_address/Forc.toml
Comment (vnepveu) [reply packages/fuels-abigen-macro/tests/test_projects/predicate_address/Forc.toml:3]: Generally I think you can try to merge all of those "small" test projects into one big one!

File: packages/fuels-abigen-macro/tests/harness.rs
Comment (vnepveu) [reply packages/fuels-abigen-macro/tests/harness.rs]: The function you mentioned is the one I had in mind, how are you blocked by feature flags? I've been using it all the time but maybe there is something I'm not aware of.

File: packages/fuels-abigen-macro/tests/harness.rs
Comment (vnepveu) [reply packages/fuels-abigen-macro/tests/harness.rs]: So our current workflow is that `unit tests` are put in the file that they are testing. And the harness is for "end to end" tests. 
For reuse of code you can look at other examples in the `harness.rs` file, basically you can skip the whole node/coins/wallet setup and do it just once at the beginning.

File: packages/fuels/tests/harness.rs
Comment (vnepveu) [inline packages/fuels/tests/harness.rs:2288]: you should do `provider = wallet.get_provider` then you can avoid doing `get_provider` all the time.

File: packages/fuels/tests/harness.rs
Comment (vnepveu) [inline packages/fuels/tests/harness.rs:2401]: you can do `wallet.get_asset_balance`

File: packages/fuels/tests/harness.rs
Comment (vnepveu) [inline packages/fuels/tests/harness.rs:2311]: wallet.get_asset_balance

File: packages/fuels/tests/harness.rs
Comment (vnepveu) [inline packages/fuels/tests/harness.rs:2286]: basically I think your best bet is to have a wallet with the predicate address inside, this way you can avoid doing the `provider` lookup everytime.

File: packages/fuels/tests/harness.rs
Comment (vnepveu) [inline packages/fuels/tests/harness.rs:2273]: returning `Result<(), Error>` should work too.

**Review by vnepveu — CHANGES_REQUESTED** [review]
Starting to look good! 
you can still streamline the process with the providers and wallets.

**Comment by hal3e** [comment]
Good work! I will add some nits from my side

**Comment by hal3e** [comment]
@camsjams sorry! I wanted to suggest a change with a commit but it was applied directly. I did it on this branch :/ You can check it out and if you like it you can leave it.

**Review by hal3e — CHANGES_REQUESTED** [review]

File: packages/fuels-contract/src/predicate.rs
Comment (MujkicA) [inline packages/fuels-contract/src/predicate.rs:26]: Use `build_transfer_tx()` from `Provider`

File: packages/fuels-contract/src/predicate.rs
Comment (MujkicA) [inline packages/fuels-contract/src/predicate.rs:57]: I dislike that we are duplicating code here from `wallet.transfer()`.

Maybe we could either move this to `wallet` and have it  be a wrapper for `wallet.transfer()` or extract `wallet.transfer()`s functionality into a provider method and have `wallet` and `predicate` use it. 
`spend_predicate` from below could also benefit from this.

This can also be done in a separate refactoring issue, just wanted to share my thoughts.

File: packages/fuels/tests/harness.rs
Comment (MujkicA) [inline packages/fuels/tests/harness.rs:2256]: The tests seem to be doing some repeated setup work. You could extract that into a separate function to avoid duplication.

File: packages/fuels/tests/harness.rs
Comment (MujkicA) [inline packages/fuels/tests/harness.rs:2261]: Type annotation seems unnecessary

File: packages/fuels/tests/harness.rs
Comment (MujkicA) [inline packages/fuels/tests/harness.rs:2272]: Annotation here too

**Review by MujkicA — COMMENTED** [review]
I've added a couple of suggestions. One thing I'm not sure about is do we need all those tests. IIUC it checks if predicates work with different types of data but the question is should that be the responsibility of the SDK vs. just confirming that we reliably transfer the data.

File: packages/fuels-contract/src/predicate.rs
Comment (hal3e) [reply packages/fuels-contract/src/predicate.rs:57]: I would suggest a separate issue that someone from the SDK team takes - I would volunteer.  We also need to add documentation as I have not seen an open issuer for that.

**Comment by simonr0204** [comment]
Is `deploy_predicate` terminology that we want to commit to? Technically nothing is deployed (and this is one of the great advantages of predicates - no state use). All predicate roots "are out" there ready to receive funds, just as all wallet addresses are (in fact, they're the same space), so nothing is deployed any more than a wallet address could be said to be when you send it coins.

**Comment by adlerjohn** [comment]
Deploy isn't the best word, for the reasons you listed. You can send funds to a predicate (hash/root).

**Comment by hal3e** [comment]
I will support @camsjams with this PR to get it closed faster :rocket: 

I did some refactoring and tried to implement the recent comments.

Changes:
- removed the `#[cfg(feature = "fuel-core-lib")]`. Now test work with and without `--all-features`
- updated how node config is used (predicates were not enabled with when running `fuel-core` bin
- `deploy_predicate` was renamed to `transfer_to_predicate` and  moved to `wallet`. It is now a wrapper of `transfer`
- `spend_predicate` was moved to `provider` with a wrapper function `recieve_from_predicate` in `wallet`
- tests were slightly refactored and I left all of them there.

**Review by iqdecay — CHANGES_REQUESTED** [review]
Few nits left

File: packages/fuels-signers/src/wallet.rs
Comment (hal3e) [reply packages/fuels-signers/src/wallet.rs:296]: Good question :smile: :. It does the same thing. The `transfer_to_predicate` used to be a method of `Predicate` were code was reused from `transfer`. An advice was to make it a wrapper of `transfer` (that is the current state). However, I also moved it to wallet as it did not make sense to keep it in `Predicate`. That said,  we could also use `transfer` directly.

On the other hand we have a wrapper for `spend_predicate` which is called `receive_from_predicate` in wallet so for me it made sense to also have `transfer_to_predicate` in wallet.

I would like to hear some opinions whether to keep the `transfer_to_predicate` function or use `transfer` directly.

Comment (iqdecay) [reply packages/fuels-signers/src/wallet.rs:296]: IMO better to just use `transfer` directly! No use exposing two different API that do the same thing :)

File: packages/fuels-signers/src/provider.rs
Comment (MujkicA) [inline packages/fuels-signers/src/provider.rs:458]: IIUC we only supporting spending, but not sending to predicates so far

**Review by MujkicA — APPROVED** [review]
We should either add docs to this PR or create another issue. Other than that, LGTM.
Great work!

**Comment by hal3e** [comment]
Here is the issue https://github.com/FuelLabs/fuels-rs/issues/535. I will add the docs after we merge this PR.

**Review by hal3e — APPROVED** [review]

**Review by Br1ght0ne — APPROVED** [review]

**Review by iqdecay — APPROVED** [review]
Nice! :ship:

**Pull request merged**
